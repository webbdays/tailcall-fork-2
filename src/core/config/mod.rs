pub use apollo::*;
pub use config::*;
pub use config_module::*;
pub use key_values::*;
pub use link::*;
pub use npo::QueryPath;
pub use reader_context::*;
pub use server::*;
pub use source::*;
pub use telemetry::*;
pub use upstream::*;
mod apollo;
mod config;
mod config_module;
pub mod cors;
mod from_document;
pub mod group_by;
mod headers;
mod into_document;
mod key_values;
mod link;
mod npo;
pub mod reader;
pub mod reader_context;
mod server;
mod source;
mod telemetry;
pub mod transformer;
mod upstream;
