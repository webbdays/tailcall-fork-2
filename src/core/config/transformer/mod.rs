mod ambiguous_type;
mod consolidate_url;
mod flatten_single_field;
mod improve_type_names;
mod merge_types;
mod nested_unions;
mod preset;
mod required;
mod tree_shake;
mod union_input_type;

pub use ambiguous_type::{AmbiguousType, Resolution};
pub use consolidate_url::ConsolidateURL;
pub use flatten_single_field::FlattenSingleField;
pub use improve_type_names::ImproveTypeNames;
pub use merge_types::TypeMerger;
pub use nested_unions::NestedUnions;
pub use preset::Preset;
pub use required::Required;
pub use tree_shake::TreeShake;
pub use union_input_type::UnionInputType;
