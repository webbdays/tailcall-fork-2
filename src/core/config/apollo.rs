use serde::{Deserialize, Serialize};

use crate::core::config::ConfigReaderContext;
use crate::core::is_default;
use crate::core::mustache::Mustache;

#[derive(Serialize, Deserialize, Clone, Debug, Default, PartialEq, Eq, schemars::JsonSchema)]
#[serde(rename_all = "camelCase")]
pub struct Apollo {
    ///
    /// Setting `apiKey` for Apollo.
    pub api_key: String,
    ///
    /// Setting `graphRef` for Apollo in the format <graphId>@<variant>.
    pub graph_ref: String,
    ///
    /// Setting `userVersion` for Apollo.
    #[serde(default, skip_serializing_if = "is_default")]
    pub user_version: Option<String>,
    ///
    /// Setting `platform` for Apollo.
    #[serde(default, skip_serializing_if = "is_default")]
    pub platform: Option<String>,
    ///
    /// Setting `version` for Apollo.
    #[serde(default, skip_serializing_if = "is_default")]
    pub version: Option<String>,
}

impl Apollo {
    pub fn render_mustache(&mut self, reader_ctx: &ConfigReaderContext) -> anyhow::Result<()> {
        let Apollo { api_key, graph_ref, user_version, platform, version } = self;

        let api_key_tmpl = Mustache::parse(api_key)?;
        *api_key = api_key_tmpl.render(reader_ctx);

        let graph_ref_tmpl = Mustache::parse(graph_ref)?;
        *graph_ref = graph_ref_tmpl.render(reader_ctx);

        let user_version_tmpl = Mustache::parse(user_version.as_deref().unwrap_or_default())?;
        *user_version = Some(user_version_tmpl.render(reader_ctx));

        let platform_tmpl = Mustache::parse(platform.as_deref().unwrap_or_default())?;
        *platform = Some(platform_tmpl.render(reader_ctx));

        let version_tmpl = Mustache::parse(version.as_deref().unwrap_or_default())?;
        *version = Some(version_tmpl.render(reader_ctx));

        Ok(())
    }
}
