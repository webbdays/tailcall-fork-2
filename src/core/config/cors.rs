use hyper::header;
use serde::{Deserialize, Serialize};

use crate::core::http::Method;
use crate::core::is_default;
use crate::core::macros::MergeRight;
use crate::core::merge_right::MergeRight;

/// Type to configure Cross-Origin Resource Sharing (CORS) for a server.
#[derive(
    Serialize, Deserialize, Clone, Debug, Default, PartialEq, Eq, schemars::JsonSchema, MergeRight,
)]
#[serde(rename_all = "camelCase")]
pub struct Cors {
    /// Indicates whether the server allows credentials (e.g., cookies,
    /// authorization headers) to be sent in cross-origin requests.
    #[serde(default, skip_serializing_if = "is_default")]
    pub allow_credentials: Option<bool>,

    /// A list of allowed headers in cross-origin requests.
    /// This can be used to specify custom headers that are allowed to be
    /// included in cross-origin requests.
    #[serde(default, skip_serializing_if = "is_default")]
    pub allow_headers: Vec<String>,

    /// A list of allowed HTTP methods in cross-origin requests.
    /// These methods specify the actions that are permitted in cross-origin
    /// requests.
    #[serde(default, skip_serializing_if = "is_default")]
    pub allow_methods: Vec<Method>,

    /// A list of origins that are allowed to access the server's resources in
    /// cross-origin requests. An origin can be a domain, a subdomain, or
    /// even 'null' for local file schemes.
    #[serde(default, skip_serializing_if = "is_default")]
    pub allow_origins: Vec<String>,

    /// Indicates whether requests from private network addresses are allowed in
    /// cross-origin requests. Private network addresses typically include
    /// IP addresses reserved for internal networks.
    #[serde(default, skip_serializing_if = "is_default")]
    pub allow_private_network: Option<bool>,

    /// A list of headers that the server exposes to the browser in cross-origin
    /// responses. Exposing certain headers allows the client-side code to
    /// access them in the response.
    #[serde(default, skip_serializing_if = "is_default")]
    pub expose_headers: Vec<String>,

    /// The maximum time (in seconds) that the client should cache preflight
    /// OPTIONS requests in order to avoid sending excessive requests to the
    /// server.
    #[serde(default, skip_serializing_if = "is_default")]
    pub max_age: Option<usize>,

    /// A list of header names that indicate the values of which might cause the
    /// server's response to vary, potentially affecting caching.
    #[serde(
        default = "preflight_request_headers",
        skip_serializing_if = "is_default"
    )]
    pub vary: Vec<String>,
}

fn preflight_request_headers() -> Vec<String> {
    vec![
        header::ORIGIN.to_string(),
        header::ACCESS_CONTROL_REQUEST_METHOD.to_string(),
        header::ACCESS_CONTROL_REQUEST_HEADERS.to_string(),
    ]
}
