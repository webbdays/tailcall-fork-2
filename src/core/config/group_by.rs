use serde::{Deserialize, Serialize};

use crate::core::is_default;

/// The `groupBy` parameter groups multiple data requests into a single call. For more details please refer out [n + 1 guide](https://tailcall.run/docs/guides/n+1#solving-using-batching).
#[derive(Clone, Debug, Eq, Serialize, Deserialize, PartialEq, schemars::JsonSchema)]
pub struct GroupBy {
    #[serde(default, skip_serializing_if = "is_default")]
    path: Vec<String>,
    #[serde(default, skip_serializing_if = "is_default")]
    key: Option<String>,
}

impl GroupBy {
    pub fn new(path: Vec<String>, key: Option<String>) -> Self {
        Self { path, key }
    }

    pub fn path(&self) -> Vec<String> {
        if self.path.is_empty() {
            return vec![String::from(ID)];
        }
        self.path.clone()
    }

    pub fn key(&self) -> &str {
        match &self.key {
            Some(value) => value,
            None => {
                if self.path.is_empty() {
                    return ID;
                }
                self.path.last().unwrap()
            }
        }
    }
}

const ID: &str = "id";

impl Default for GroupBy {
    fn default() -> Self {
        Self { path: vec![ID.to_string()], key: None }
    }
}
