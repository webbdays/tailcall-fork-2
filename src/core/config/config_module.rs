use std::collections::{HashMap, HashSet};
use std::ops::Deref;
use std::sync::Arc;

use jsonwebtoken::jwk::JwkSet;
use prost_reflect::prost_types::{FileDescriptorProto, FileDescriptorSet};
use rustls_pki_types::{CertificateDer, PrivateKeyDer};

use crate::core::config::Config;
use crate::core::macros::MergeRight;
use crate::core::merge_right::MergeRight;
use crate::core::proto_reader::ProtoMetadata;
use crate::core::rest::{EndpointSet, Unchecked};
use crate::core::valid::{Valid, Validator};
use crate::core::Transform;

/// A wrapper on top of Config that contains all the resolved extensions and
/// computed values.
#[derive(Clone, Debug, Default, MergeRight)]
pub struct ConfigModule {
    extensions: Extensions,
    cache: Cache,
}

/// A cache that store resolved input, output and interface types so that it's
/// not computed again and again.
#[derive(Clone, Debug, Default)]
struct Cache {
    config: Config,
    input_types: HashSet<String>,
    output_types: HashSet<String>,
    interface_types: HashSet<String>,
}

impl From<Config> for Cache {
    fn from(value: Config) -> Self {
        let input_types = value.input_types();
        let output_types = value.output_types();
        let interface_types = value.interface_types();

        Cache {
            config: value,
            input_types: input_types.clone(),
            output_types: output_types.clone(),
            interface_types: interface_types.clone(),
        }
    }
}

impl MergeRight for Cache {
    fn merge_right(self, other: Self) -> Self {
        Cache::from(self.config.merge_right(other.config))
    }
}

impl ConfigModule {
    pub fn new(config: Config, extensions: Extensions) -> Self {
        ConfigModule { cache: Cache::from(config), extensions }
    }

    pub fn set_extensions(mut self, extensions: Extensions) -> Self {
        self.extensions = extensions;
        self
    }

    pub fn merge_extensions(mut self, extensions: Extensions) -> Self {
        self.extensions = self.extensions.merge_right(extensions);
        self
    }

    pub fn config(&self) -> &Config {
        &self.cache.config
    }

    pub fn extensions(&self) -> &Extensions {
        &self.extensions
    }

    pub fn input_types(&self) -> &HashSet<String> {
        &self.cache.input_types
    }

    pub fn output_types(&self) -> &HashSet<String> {
        &self.cache.output_types
    }

    pub fn interface_types(&self) -> &HashSet<String> {
        &self.cache.interface_types
    }

    pub fn transform<T: Transform<Value = Config>>(self, transformer: T) -> Valid<Self, T::Error> {
        transformer
            .transform(self.cache.config)
            .map(|config| ConfigModule::new(config, self.extensions))
    }
}

#[derive(Clone, Debug, Default)]
pub struct Content<A> {
    pub id: Option<String>,
    pub content: A,
}

impl<A> Deref for Content<A> {
    type Target = A;
    fn deref(&self) -> &Self::Target {
        &self.content
    }
}

/// Extensions are meta-information required before we can generate the
/// blueprint. Typically, this information cannot be inferred without performing
/// an IO operation, i.e., reading a file, making an HTTP call, etc.
#[derive(Clone, Debug, Default, MergeRight)]
pub struct Extensions {
    /// Contains the file descriptor set resolved from the links to proto files
    pub grpc_file_descriptors: HashMap<String, FileDescriptorProto>,

    /// Contains the contents of the JS file
    pub script: Option<String>,

    /// Contains the certificate used on HTTP2 with TLS
    pub cert: Vec<CertificateDer<'static>>,

    /// Contains the key used on HTTP2 with TLS
    pub keys: Arc<Vec<PrivateKeyDer<'static>>>,

    /// Contains the endpoints
    pub endpoint_set: EndpointSet<Unchecked>,

    pub htpasswd: Vec<Content<String>>,

    pub jwks: Vec<Content<JwkSet>>,
}

impl Extensions {
    pub fn add_proto(&mut self, metadata: ProtoMetadata) {
        for file in metadata.descriptor_set.file {
            self.grpc_file_descriptors
                .insert(file.name().to_string(), file);
        }
    }

    pub fn get_file_descriptor_set(&self) -> FileDescriptorSet {
        FileDescriptorSet { file: self.grpc_file_descriptors.values().cloned().collect() }
    }

    pub fn has_auth(&self) -> bool {
        !self.htpasswd.is_empty() || !self.jwks.is_empty()
    }
}

impl MergeRight for FileDescriptorSet {
    fn merge_right(mut self, other: Self) -> Self {
        self.file.extend(other.file);

        self
    }
}

impl Deref for ConfigModule {
    type Target = Config;
    fn deref(&self) -> &Self::Target {
        self.config()
    }
}

impl From<Config> for ConfigModule {
    fn from(config: Config) -> Self {
        ConfigModule { cache: Cache::from(config), ..Default::default() }
    }
}
