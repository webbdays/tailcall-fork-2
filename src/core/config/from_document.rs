use std::collections::{BTreeMap, BTreeSet};

use async_graphql::parser::types::{
    BaseType, ConstDirective, EnumType, FieldDefinition, InputObjectType, InputValueDefinition,
    InterfaceType, ObjectType, SchemaDefinition, ServiceDocument, Type, TypeDefinition, TypeKind,
    TypeSystemDefinition, UnionType,
};
use async_graphql::parser::Positioned;
use async_graphql::Name;
use async_graphql_value::ConstValue;

use super::telemetry::Telemetry;
use super::{Alias, Tag, JS};
use crate::core::config::{
    self, Cache, Call, Config, Enum, GraphQL, Grpc, Link, Modify, Omit, Protected, RootSchema,
    Server, Union, Upstream, Variant,
};
use crate::core::directive::DirectiveCodec;
use crate::core::valid::{Valid, ValidationError, Validator};

const DEFAULT_SCHEMA_DEFINITION: &SchemaDefinition = &SchemaDefinition {
    extend: false,
    directives: Vec::new(),
    query: None,
    mutation: None,
    subscription: None,
};

pub fn from_document(doc: ServiceDocument) -> Valid<Config, String> {
    let type_definitions: Vec<_> = doc
        .definitions
        .iter()
        .filter_map(|def| match def {
            TypeSystemDefinition::Type(td) => Some(td),
            _ => None,
        })
        .collect();

    let types = to_types(&type_definitions);
    let unions = to_union_types(&type_definitions);
    let enums = to_enum_types(&type_definitions);
    let schema = schema_definition(&doc).map(to_root_schema);
    schema_definition(&doc).and_then(|sd| {
        server(sd)
            .fuse(upstream(sd))
            .fuse(types)
            .fuse(unions)
            .fuse(enums)
            .fuse(schema)
            .fuse(links(sd))
            .fuse(telemetry(sd))
            .map(
                |(server, upstream, types, unions, enums, schema, links, telemetry)| Config {
                    server,
                    upstream,
                    types,
                    unions,
                    enums,
                    schema,
                    links,
                    telemetry,
                },
            )
    })
}

fn schema_definition(doc: &ServiceDocument) -> Valid<&SchemaDefinition, String> {
    doc.definitions
        .iter()
        .find_map(|def| match def {
            TypeSystemDefinition::Schema(schema_definition) => Some(&schema_definition.node),
            _ => None,
        })
        .map_or_else(|| Valid::succeed(DEFAULT_SCHEMA_DEFINITION), Valid::succeed)
}

fn process_schema_directives<T: DirectiveCodec<T> + Default>(
    schema_definition: &SchemaDefinition,
    directive_name: &str,
) -> Valid<T, String> {
    let mut res = Valid::succeed(T::default());
    for directive in schema_definition.directives.iter() {
        if directive.node.name.node.as_ref() == directive_name {
            res = T::from_directive(&directive.node);
        }
    }
    res
}

fn process_schema_multiple_directives<T: DirectiveCodec<T> + Default>(
    schema_definition: &SchemaDefinition,
    directive_name: &str,
) -> Valid<Vec<T>, String> {
    let directives: Vec<Valid<T, String>> = schema_definition
        .directives
        .iter()
        .filter_map(|directive| {
            if directive.node.name.node.as_ref() == directive_name {
                Some(T::from_directive(&directive.node))
            } else {
                None
            }
        })
        .collect();

    Valid::from_iter(directives, |item| item)
}

fn server(schema_definition: &SchemaDefinition) -> Valid<Server, String> {
    process_schema_directives(schema_definition, config::Server::directive_name().as_str())
}

fn upstream(schema_definition: &SchemaDefinition) -> Valid<Upstream, String> {
    process_schema_directives(
        schema_definition,
        config::Upstream::directive_name().as_str(),
    )
}

fn links(schema_definition: &SchemaDefinition) -> Valid<Vec<Link>, String> {
    process_schema_multiple_directives(schema_definition, config::Link::directive_name().as_str())
}

fn telemetry(schema_definition: &SchemaDefinition) -> Valid<Telemetry, String> {
    process_schema_directives(
        schema_definition,
        config::telemetry::Telemetry::directive_name().as_str(),
    )
}

fn to_root_schema(schema_definition: &SchemaDefinition) -> RootSchema {
    let query = schema_definition.query.as_ref().map(pos_name_to_string);
    let mutation = schema_definition.mutation.as_ref().map(pos_name_to_string);
    let subscription = schema_definition
        .subscription
        .as_ref()
        .map(pos_name_to_string);

    RootSchema { query, mutation, subscription }
}
fn pos_name_to_string(pos: &Positioned<Name>) -> String {
    pos.node.to_string()
}
fn to_types(
    type_definitions: &Vec<&Positioned<TypeDefinition>>,
) -> Valid<BTreeMap<String, config::Type>, String> {
    Valid::from_iter(type_definitions, |type_definition| {
        let type_name = pos_name_to_string(&type_definition.node.name);
        match type_definition.node.kind.clone() {
            TypeKind::Object(object_type) => to_object_type(
                &object_type,
                &type_definition.node.description,
                &type_definition.node.directives,
            )
            .some(),
            TypeKind::Interface(interface_type) => to_object_type(
                &interface_type,
                &type_definition.node.description,
                &type_definition.node.directives,
            )
            .some(),
            TypeKind::Enum(_) => Valid::none(),
            TypeKind::InputObject(input_object_type) => to_input_object(
                input_object_type,
                &type_definition.node.description,
                &type_definition.node.directives,
            )
            .some(),
            TypeKind::Union(_) => Valid::none(),
            TypeKind::Scalar => Valid::succeed(Some(to_scalar_type())),
        }
        .map(|option| (type_name, option))
    })
    .map(|vec| {
        BTreeMap::from_iter(
            vec.into_iter()
                .filter_map(|(name, option)| option.map(|tpe| (name, tpe))),
        )
    })
}
fn to_scalar_type() -> config::Type {
    config::Type { ..Default::default() }
}
fn to_union_types(
    type_definitions: &[&Positioned<TypeDefinition>],
) -> Valid<BTreeMap<String, Union>, String> {
    Valid::succeed(
        type_definitions
            .iter()
            .filter_map(|type_definition| {
                let type_name = pos_name_to_string(&type_definition.node.name);
                let type_opt = match type_definition.node.kind.clone() {
                    TypeKind::Union(union_type) => to_union(
                        union_type,
                        &type_definition
                            .node
                            .description
                            .to_owned()
                            .map(|pos| pos.node),
                    ),
                    _ => return None,
                };
                Some((type_name, type_opt))
            })
            .collect(),
    )
}

fn to_enum_types(
    type_definitions: &[&Positioned<TypeDefinition>],
) -> Valid<BTreeMap<String, Enum>, String> {
    Valid::from_iter(type_definitions.iter(), |type_definition| {
        let type_name = pos_name_to_string(&type_definition.node.name);
        let type_opt = match type_definition.node.kind.clone() {
            TypeKind::Enum(enum_type) => to_enum(
                enum_type,
                type_definition
                    .node
                    .description
                    .to_owned()
                    .map(|pos| pos.node),
            ),
            _ => return Valid::succeed(None),
        };
        type_opt.map(|type_opt| Some((type_name, type_opt)))
    })
    .map(|values| values.into_iter().flatten().collect())
}

#[allow(clippy::too_many_arguments)]
fn to_object_type<T>(
    object: &T,
    description: &Option<Positioned<String>>,
    directives: &[Positioned<ConstDirective>],
) -> Valid<config::Type, String>
where
    T: ObjectLike,
{
    let fields = object.fields();
    let implements = object.implements();

    Cache::from_directives(directives.iter())
        .fuse(to_fields(fields))
        .fuse(Protected::from_directives(directives.iter()))
        .fuse(Tag::from_directives(directives.iter()))
        .map(|(cache, fields, protected, tag)| {
            let doc = description.to_owned().map(|pos| pos.node);
            let implements = implements.iter().map(|pos| pos.node.to_string()).collect();
            let added_fields = to_add_fields_from_directives(directives);
            config::Type { fields, added_fields, doc, implements, cache, protected, tag }
        })
}
fn to_input_object(
    input_object_type: InputObjectType,
    description: &Option<Positioned<String>>,
    directives: &[Positioned<ConstDirective>],
) -> Valid<config::Type, String> {
    to_input_object_fields(&input_object_type.fields)
        .fuse(Protected::from_directives(directives.iter()))
        .map(|(fields, protected)| {
            let doc = description.to_owned().map(|pos| pos.node);
            config::Type { fields, protected, doc, ..Default::default() }
        })
}

fn to_fields_inner<T, F>(
    fields: &Vec<Positioned<T>>,
    transform: F,
) -> Valid<BTreeMap<String, config::Field>, String>
where
    F: Fn(&T) -> Valid<config::Field, String>,
    T: HasName,
{
    Valid::from_iter(fields, |field| {
        let field_name = pos_name_to_string(field.node.name());
        transform(&field.node).map(|field| (field_name, field))
    })
    .map(BTreeMap::from_iter)
}
fn to_fields(
    fields: &Vec<Positioned<FieldDefinition>>,
) -> Valid<BTreeMap<String, config::Field>, String> {
    to_fields_inner(fields, to_field)
}
fn to_input_object_fields(
    input_object_fields: &Vec<Positioned<InputValueDefinition>>,
) -> Valid<BTreeMap<String, config::Field>, String> {
    to_fields_inner(input_object_fields, to_input_object_field)
}
fn to_field(field_definition: &FieldDefinition) -> Valid<config::Field, String> {
    to_common_field(field_definition, to_args(field_definition), None)
}
fn to_input_object_field(field_definition: &InputValueDefinition) -> Valid<config::Field, String> {
    to_common_field(
        field_definition,
        BTreeMap::new(),
        field_definition
            .default_value
            .as_ref()
            .map(|f| f.node.clone()),
    )
}
fn to_common_field<F>(
    field: &F,
    args: BTreeMap<String, config::Arg>,
    default_value: Option<ConstValue>,
) -> Valid<config::Field, String>
where
    F: FieldLike,
{
    let type_of = field.type_of();
    let base = &type_of.base;
    let nullable = &type_of.nullable;
    let description = field.description();
    let directives = field.directives();
    let default_value = default_value
        .map(ConstValue::into_json)
        .transpose()
        .map_err(|err| ValidationError::new(err.to_string()))
        .into();

    let type_of = to_type_of(type_of);
    let list = matches!(&base, BaseType::List(_));
    let list_type_required = matches!(&base, BaseType::List(type_of) if !type_of.nullable);
    let doc = description.to_owned().map(|pos| pos.node);
    config::Http::from_directives(directives.iter())
        .fuse(GraphQL::from_directives(directives.iter()))
        .fuse(Cache::from_directives(directives.iter()))
        .fuse(Grpc::from_directives(directives.iter()))
        .fuse(Omit::from_directives(directives.iter()))
        .fuse(Modify::from_directives(directives.iter()))
        .fuse(JS::from_directives(directives.iter()))
        .fuse(Call::from_directives(directives.iter()))
        .fuse(Protected::from_directives(directives.iter()))
        .fuse(default_value)
        .map(
            |(http, graphql, cache, grpc, omit, modify, script, call, protected, default_value)| {
                let const_field = to_const_field(directives);
                config::Field {
                    type_of,
                    list,
                    required: !nullable,
                    list_type_required,
                    args,
                    doc,
                    modify,
                    omit,
                    http,
                    grpc,
                    script,
                    const_field,
                    graphql,
                    cache,
                    call,
                    protected,
                    default_value,
                }
            },
        )
}

fn to_type_of(type_: &Type) -> String {
    match &type_.base {
        BaseType::Named(name) => name.to_string(),
        BaseType::List(ty) => to_type_of(ty),
    }
}
fn to_args(field_definition: &FieldDefinition) -> BTreeMap<String, config::Arg> {
    let mut args: BTreeMap<String, config::Arg> = BTreeMap::new();

    for arg in field_definition.arguments.iter() {
        let arg_name = pos_name_to_string(&arg.node.name);
        let arg_val = to_arg(&arg.node);
        args.insert(arg_name, arg_val);
    }

    args
}
fn to_arg(input_value_definition: &InputValueDefinition) -> config::Arg {
    let type_of = to_type_of(&input_value_definition.ty.node);
    let list = matches!(&input_value_definition.ty.node.base, BaseType::List(_));
    let required = !input_value_definition.ty.node.nullable;
    let doc = input_value_definition
        .description
        .to_owned()
        .map(|pos| pos.node);
    let modify = Modify::from_directives(input_value_definition.directives.iter())
        .to_result()
        .ok()
        .flatten();
    let default_value = if let Some(pos) = input_value_definition.default_value.as_ref() {
        let value = &pos.node;
        serde_json::to_value(value).ok()
    } else {
        None
    };
    config::Arg { type_of, list, required, doc, modify, default_value }
}

fn to_union(union_type: UnionType, doc: &Option<String>) -> Union {
    let types = union_type
        .members
        .iter()
        .map(|member| member.node.to_string())
        .collect();
    Union { types, doc: doc.clone() }
}

fn to_enum(enum_type: EnumType, doc: Option<String>) -> Valid<Enum, String> {
    let variants = Valid::from_iter(enum_type.values.iter(), |member| {
        let name = member.node.value.node.as_str().to_owned();
        let alias = member
            .node
            .directives
            .iter()
            .find(|d| d.node.name.node.as_str() == Alias::directive_name());
        if let Some(alias) = alias {
            Alias::from_directive(&alias.node).map(|alias| Variant { name, alias: Some(alias) })
        } else {
            Valid::succeed(Variant { name, alias: None })
        }
    });
    variants.map(|v| Enum { variants: v.into_iter().collect::<BTreeSet<Variant>>(), doc })
}
fn to_const_field(directives: &[Positioned<ConstDirective>]) -> Option<config::Expr> {
    directives.iter().find_map(|directive| {
        if directive.node.name.node == config::Expr::directive_name() {
            config::Expr::from_directive(&directive.node)
                .to_result()
                .ok()
        } else {
            None
        }
    })
}

fn to_add_fields_from_directives(
    directives: &[Positioned<ConstDirective>],
) -> Vec<config::AddField> {
    directives
        .iter()
        .filter_map(|directive| {
            if directive.node.name.node == config::AddField::directive_name() {
                config::AddField::from_directive(&directive.node)
                    .to_result()
                    .ok()
            } else {
                None
            }
        })
        .collect::<Vec<_>>()
}

trait HasName {
    fn name(&self) -> &Positioned<Name>;
}
impl HasName for FieldDefinition {
    fn name(&self) -> &Positioned<Name> {
        &self.name
    }
}
impl HasName for InputValueDefinition {
    fn name(&self) -> &Positioned<Name> {
        &self.name
    }
}

trait FieldLike {
    fn type_of(&self) -> &Type;
    fn description(&self) -> &Option<Positioned<String>>;
    fn directives(&self) -> &[Positioned<ConstDirective>];
}
impl FieldLike for FieldDefinition {
    fn type_of(&self) -> &Type {
        &self.ty.node
    }
    fn description(&self) -> &Option<Positioned<String>> {
        &self.description
    }
    fn directives(&self) -> &[Positioned<ConstDirective>] {
        &self.directives
    }
}
impl FieldLike for InputValueDefinition {
    fn type_of(&self) -> &Type {
        &self.ty.node
    }
    fn description(&self) -> &Option<Positioned<String>> {
        &self.description
    }
    fn directives(&self) -> &[Positioned<ConstDirective>] {
        &self.directives
    }
}

trait ObjectLike {
    fn fields(&self) -> &Vec<Positioned<FieldDefinition>>;
    fn implements(&self) -> &Vec<Positioned<Name>>;
}
impl ObjectLike for ObjectType {
    fn fields(&self) -> &Vec<Positioned<FieldDefinition>> {
        &self.fields
    }
    fn implements(&self) -> &Vec<Positioned<Name>> {
        &self.implements
    }
}
impl ObjectLike for InterfaceType {
    fn fields(&self) -> &Vec<Positioned<FieldDefinition>> {
        &self.fields
    }
    fn implements(&self) -> &Vec<Positioned<Name>> {
        &self.implements
    }
}
