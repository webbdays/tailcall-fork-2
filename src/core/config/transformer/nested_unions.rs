use std::collections::{BTreeMap, BTreeSet, HashSet};

use crate::core::config::{Config, Union};
use crate::core::transform::Transform;
use crate::core::valid::{Valid, Validator};

/// Transforms unions by replacing each nested union in union definition
/// recursively by their actual types
#[derive(Default)]
pub struct NestedUnions;

impl Transform for NestedUnions {
    type Value = Config;
    type Error = String;

    fn transform(&self, mut config: Config) -> Valid<Config, String> {
        let visitor = Visitor { unions: &config.unions };

        visitor.visit().map(|unions| {
            config.unions = unions;
            config
        })
    }
}

struct Visitor<'cfg> {
    unions: &'cfg BTreeMap<String, Union>,
}

impl<'cfg> Visitor<'cfg> {
    fn visit(self) -> Valid<BTreeMap<String, Union>, String> {
        let mut result = BTreeMap::new();

        Valid::from_iter(self.unions.iter(), |(union_name, union_)| {
            let mut union_types = BTreeSet::new();

            self.walk_union(union_, &mut union_types, &mut HashSet::new())
                .trace(union_name)
                .map(|_| {
                    let new_union = Union { types: union_types, ..union_.clone() };

                    result.insert(union_name.clone(), new_union);
                })
        })
        .map(|_| result)
    }

    fn walk_union(
        &'cfg self,
        union_: &'cfg Union,
        union_types: &mut BTreeSet<String>,
        seen: &mut HashSet<&'cfg String>,
    ) -> Valid<(), String> {
        Valid::from_iter(union_.types.iter(), |type_name| {
            if let Some(union_) = self.unions.get(type_name) {
                if seen.contains(type_name) {
                    return Valid::fail(format!("Recursive type {type_name}"));
                }

                seen.insert(type_name);
                self.walk_union(union_, union_types, seen)
            } else {
                union_types.insert(type_name.clone());
                Valid::succeed(())
            }
        })
        .unit()
    }
}

#[cfg(test)]
mod tests {
    use insta::assert_snapshot;

    use super::NestedUnions;
    use crate::core::config::Config;
    use crate::core::transform::Transform;
    use crate::core::valid::Validator;

    #[test]
    fn test_nested_unions() {
        let config =
            std::fs::read_to_string(tailcall_fixtures::configs::YAML_NESTED_UNIONS).unwrap();
        let config = Config::from_yaml(&config).unwrap();
        let config = NestedUnions.transform(config).to_result().unwrap();

        assert_snapshot!(config.to_sdl());
    }

    #[test]
    fn test_nested_unions_recursive() {
        let config =
            std::fs::read_to_string(tailcall_fixtures::configs::YAML_NESTED_UNIONS_RECURSIVE)
                .unwrap();
        let config = Config::from_yaml(&config).unwrap();
        let error = NestedUnions.transform(config).to_result().unwrap_err();

        assert_snapshot!(error);
    }
}
