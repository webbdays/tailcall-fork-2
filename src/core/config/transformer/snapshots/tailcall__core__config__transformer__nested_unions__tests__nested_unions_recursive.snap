---
source: src/core/config/transformer/nested_unions.rs
expression: error
---
Validation Error
• Recursive type U1 [U]
• Recursive type U2 [U]
• Recursive type U2 [U1]
• Recursive type U1 [U2]
